### PR TITLE
Refactored Parameter Validation and Unified TiledDataset Class

### PR DESCRIPTION
**Pull request by TibbersHao**
This PR covers changes as a reflection to the 2nd Diamond Hackathon's discussion.

### Feature Upgrades

- [x] Refactored the Parameter Class for Pydantic BaseModel to lift the nested structure from last version, thanks to @dylanmcreynolds !
- [x] I/O: Now both `train.py` and `segment.py` will only take one argument `yaml_file_path`.
- [x] Tiled Dataset: The `InferenceDataset` has been taken out and `TrainingDataset` has been modified to a unified `TiledDataset` which will handle both training and inference.
- [x] Testing: Example yaml files have been provided as a template under a new directory.

### To Test:

1. Modify the `tiled_api_key` for all clients under each example yaml file.
2. Modify the `uid` if necessary.
3. In terminal, `make train_tunet`
4. In terminal, `make segment_tunet`

### CODE REVIEW

**Review by dylanmcreynolds — COMMENTED** [review]
Looks good. Just one small change requested.

**Review by taxe10 — CHANGES_REQUESTED** [review]
This looks great! I left some minor comments

File: src/tiled_dataset.py [hunk 1]
```diff
@@ -25,59 +25,27 @@ def __init__(
...
-            transform:           callable, if not given return PIL image
-
-        Return:
-            ml_data:             data_tensor
-        '''
-        self.data_client = from_uri(data_tiled_uri, api_key=data_tiled_api_key)
-        self.mask_client = from_uri(mask_tiled_uri, api_key=mask_tiled_api_key)
-        self.seg_client = from_uri(seg_tiled_uri, api_key=seg_tiled_api_key)
+            self.mask_client = None
         self.mask_idx = list(mask_idx)
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/tiled_dataset.py:32]: Thanks for the check. I like the approach suggested above and will make modifications. 

Also a side note to this: the `mask_idx` is served as a temporary solution for the case of clay dataset in SPIN, and essentially this parameter, together with the `shift` below, will be taken out and modified according to an upcoming PR from seg app about mask writing.

File: example_yamls/example_tunet.yaml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# I/O
+data_tiled_uri: https://tiled-seg.als.lbl.gov/api/v1/metadata/reconstruction/rec20190524_085542_clay_testZMQ_8bit/20190524_085542_clay_testZMQ_
+data_tiled_api_key: <secret>
+mask_tiled_uri: https://tiled-seg.als.lbl.gov/api/v1/metadata/reconstruction/seg-partial-rec20190524_085542_clay_testZMQ_8bit/seg-partial-20190524_085542_clay_testZMQ_
+mask_tiled_api_key: <secret>
+seg_tiled_uri: http://0.0.0.0:8888
+seg_tiled_api_key: <secret>
+
+mask_idx: [10, 201, 222, 493]
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply example_yamls/example_tunet.yaml:11]: As mentioned above for Dylan's comment, the `mask_idx` and `shift` will be taken out according to an upcoming PR from the seg app for mask writing. So the app will not send the mask indexes as a parameter, and if I recall correctly, it will be a part as the metadata from tiled.

Please correct me if I am wrong @Wiebke .

File: example_yamls/example_tunet.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+shift: 2
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply example_yamls/example_tunet.yaml:12]: Nope, same as `mask_idx` it will soon be taken out.

File: src/parameters.py [hunk 4]
```diff
@@ -1,124 +1,51 @@
...
+    shuffle_train: Optional[bool] = Field(default=True, description="whether to shuffle data in train set")
+    batch_size_train: Optional[int] = Field(default=1, description="batch size of train set")
 
-class Normalization(str, Enum):
-    batchnorm2d = "BatchNorm2d"
-    batchnorm3d = "BatchNorm3d"
+    shuffle_val: Optional[bool] = Field(default=True, description="whether to shuffle data in validation set")
+    batch_size_val: Optional[int] = Field(default=1, description="batch size of validation set")
 
+    shuffle_inference: Optional[bool] = Field(default=False, description="whether to shuffle data during inference")
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/parameters.py:25]: I agree on the inference side, as we want the result been produced in order. However, I believe having shuffle in validation will still be giving some merit in terms of reducing the model to learn from certain patterns (from a validation perspective)?

Happy to discuss more about this before I make modifications.

File: src/parameters.py [hunk 5]
```diff
@@ -1,124 +1,51 @@
...
+    max_dilation: Optional[int] = Field(default=5, description="maximum dilation for MSDNet")
+    dilation_array: Optional[List[int]] = Field(default=None, description="customized dilation array for MSDNet")
 
-class TUNetParameters(BaseModel):    
-    depth: Optional[int] = Field(None, description='the depth of the UNet')
-    base_channels: Optional[int] = Field(None, description='the number of initial channels for UNet')
-    growth_rate: Optional[int] = Field(None, description='multiplicative growth factor of number of '\
+class TUNetParameters(TrainingParameters):    
+    depth: Optional[int] = Field(default=4, description='the depth of the UNet')
+    base_channels: Optional[int] = Field(default=8, description='the number of initial channels for UNet')
```
[the remark this answers is not in the document]
Comment (TibbersHao) [reply src/parameters.py:43]: Great spot. I was changing this to 8 to speed up the testing, and put that number there without thinking further.

Copying from DLSIA's tutorial for reference:
"Recommended parameters are depth = 4, 5, or 6; base_channels = 32 or 64; growth_rate between 1.5 and 2.5; and hidden_rate = 1"

New default value will come as below:
depth = 4
base_channels = 32
growth_rate = 2
hidden_rate = 1

for TUNet3+ the carryover_channels will be modified to 32 as default value.

File: src/parameters.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (taxe10) [reply src/parameters.py:25]: I think shuffling the validation set could be problematic as well. Since changes in the validation accuracy could change because of shuffling instead of overall model performance (since validation loss is calculated per batch and then averaged).
This can be updated in the next pull request if preferred.

**Review by taxe10 — APPROVED** [review]

File: src/tiled_dataset.py [hunk 6]
```diff
@@ -25,59 +25,27 @@ def __init__(
...
+            mask = self.mask_client[idx,].astype('int') - self.shift # Conversion to int is needed as switching unlabeled pixels to -1 would cause trouble in uint8 format
+            if self.transform:
+                return self.transform(data), mask
+            else:
+                return data, mask
         else:
-            return data
\ No newline at end of file
+            if self.transform:
+                return self.transform(data)
```
Comment (Wiebke) [inline src/tiled_dataset.py:49]: Just to double check, this is also how DLSIA's internal inference methods that we may use later on expect data to be passed (i.e. not still a tuple)?

**Review by Wiebke — APPROVED** [review]
Looks great! 
I had a small question on the refactor of the data loader, but this would be part of a future PR anyways.

File: src/tiled_dataset.py (hunk 6 above)
Comment (TibbersHao) [reply src/tiled_dataset.py:49]: As discussed in person we are good for this part as the inference function only asks for the image part, not the pair.
